lib/util.js

```diff
@@ -96,10 +96,38 @@ function isValidNumber(value) {
     return (typeof value === "number") && !isNaN(value);
 }
 
+/**
+ * Converts the content of a readable Stream into a Buffer
+ * instance. WARNING: this method will read the entirety of
+ * the stream into memory; use with extreme caution.
+ *
+ * @param {ReadableStream} stream Stream to convert.
+ * @returns {Promise<Buffer>} Contents of the stream in a
+ *  Buffer.
+ */
+async function streamToBuffer(stream) {
+    return new Promise((resolve, reject) => {
+        const chunks = [];
+
+        stream.once('error', (err) => {
+            reject(err);
+        });
+
+        stream.once('end', () => {
+            resolve(Buffer.concat(chunks));
+        });
+
+        stream.on('data', (chunk) => {
+            chunks.push(chunk);
+        });
+    });
+}
```

Comment:
This can then be moved to `transfer.js`.

Reply:
The chunks that Node.js gives us are only 16KB in size, so there will be a lot of chunks in the array. It is more efficient to call `Buffer.allocUnsafe` the buffer up-fron (since we know the exact amount of bytes we will download) and then use `chunk.copy(buffer, bufferOffset, 0)` to copy each chunk in to the buffer.

At the end for safety you can then do `buffer.slice(0, bufferOffset)` to make sure it doesn't return unwritten bytes.

Reply:
On that last note, we also need to check whether the actual amount of bytes read matches the expected bytes. If they don't we will need to retry. We've seen downloads where the stream was just closed mid way through.

Reply:
So you want this method in `Transfer` instead? Do you want it as a class function or as a function defined before the class? If the latter, I'd argue for keeping it where it is, since IMO it's more clean to have functions like this in separate files rather than mixing class and function definitions in a single file.